File: AppCheckCore/Tests/Unit/DeviceCheckProvider/GACDeviceCheckAPIServiceTests.m
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import <XCTest/XCTest.h>

#import <OCMock/OCMock.h>
#import "FBLPromise+Testing.h"

#import <GoogleUtilities/GULURLSessionDataResponse.h>

#import "AppCheckCore/Sources/Core/APIService/GACAppCheckAPIService.h"
#import "AppCheckCore/Sources/Core/Errors/GACAppCheckErrorUtil.h"
#import "AppCheckCore/Sources/DeviceCheckProvider/API/GACDeviceCheckAPIService.h"
#import "AppCheckCore/Sources/Public/AppCheckCore/GACAppCheckErrors.h"
#import "AppCheckCore/Sources/Public/AppCheckCore/GACAppCheckToken.h"

#import "AppCheckCore/Tests/Unit/Utils/GACFixtureLoader.h"
#import "AppCheckCore/Tests/Utils/URLSession/GACURLSessionOCMockStub.h"

static NSString *const kResourceName = @"projects/project_id/apps/app_id";

typedef BOOL (^FIRRequestValidationBlock)(NSURLRequest *request);

@interface GACDeviceCheckAPIServiceTests : XCTestCase
@property(nonatomic) GACDeviceCheckAPIService *APIService;

@property(nonatomic) id mockAPIService;

@end

@implementation GACDeviceCheckAPIServiceTests

- (void)setUp {
  [super setUp];

  self.mockAPIService = OCMProtocolMock(@protocol(GACAppCheckAPIServiceProtocol));
  OCMStub([self.mockAPIService baseURL]).andReturn(@"https://test.appcheck.url.com/alpha");

  self.APIService = [[GACDeviceCheckAPIService alloc] initWithAPIService:self.mockAPIService
                                                            resourceName:kResourceName];
}

- (void)tearDown {
  self.APIService = nil;
  [self.mockAPIService stopMocking];
  self.mockAPIService = nil;

  [super tearDown];
}

- (void)testAppCheckTokenSuccess {
  [self testAppCheckTokenSuccessWithLimitedUse:NO];
}

- (void)testAppCheckTokenSuccessWithLimitedUse {
  [self testAppCheckTokenSuccessWithLimitedUse:YES];
}

- (void)testAppCheckTokenSuccessWithLimitedUse:(BOOL)limitedUse {
  NSData *deviceTokenData = [@"device_token" dataUsingEncoding:NSUTF8StringEncoding];
  GACAppCheckToken *expectedResult = [[GACAppCheckToken alloc] initWithToken:@"app_check_token"
                                                              expirationDate:[NSDate date]];

  // 1. Stub API service.
  // 1.1 Stub send request.
  NSString *expectedRequestURL =
      [NSString stringWithFormat:@"%@%@", [self.mockAPIService baseURL],
                                 @"/projects/project_id/apps/app_id:exchangeDeviceCheckToken"];
  id URLValidationArg = [OCMArg checkWithBlock:^BOOL(NSURL *URL) {
    XCTAssertEqualObjects(URL.absoluteString, expectedRequestURL);
    return YES;
  }];

  id HTTPBodyValidationArg = [self HTTPBodyValidationArgWithDeviceToken:deviceTokenData
                                                             limitedUse:limitedUse];

  NSData *responseBody =
      [GACFixtureLoader loadFixtureNamed:@"FACTokenExchangeResponseSuccess.json"];
  XCTAssertNotNil(responseBody);

  NSHTTPURLResponse *HTTPResponse = [GACURLSessionOCMockStub HTTPResponseWithCode:200];
  GULURLSessionDataResponse *APIResponse =
      [[GULURLSessionDataResponse alloc] initWithResponse:HTTPResponse HTTPBody:responseBody];

  OCMExpect([self.mockAPIService sendRequestWithURL:URLValidationArg
                                         HTTPMethod:@"POST"
                                               body:HTTPBodyValidationArg
                                  additionalHeaders:@{@"Content-Type" : @"application/json"}])
      .andReturn([FBLPromise resolvedWith:APIResponse]);

  // 1.2. Stub response parsing.
  OCMExpect([self.mockAPIService appCheckTokenWithAPIResponse:APIResponse])
      .andReturn([FBLPromise resolvedWith:expectedResult]);

  // 2. Send request.
  __auto_type tokenPromise = [self.APIService appCheckTokenWithDeviceToken:deviceTokenData
                                                                limitedUse:limitedUse];

  // 3. Verify.
  XCTAssert(FBLWaitForPromisesWithTimeout(1));

  XCTAssertTrue(tokenPromise.isFulfilled);
  XCTAssertNil(tokenPromise.error);

  XCTAssertEqualObjects(tokenPromise.value.token, expectedResult.token);
  XCTAssertEqualObjects(tokenPromise.value.expirationDate, expectedResult.expirationDate);

  OCMVerifyAll(self.mockAPIService);
}

- (void)testAppCheckTokenResponseParsingError {
  NSData *deviceTokenData = [@"device_token" dataUsingEncoding:NSUTF8StringEncoding];
  NSError *parsingError = [NSError errorWithDomain:@"testAppCheckTokenResponseParsingError"
                                              code:-1
                                          userInfo:nil];

  // 1. Stub API service.
  // 1.1 Stub send request.
  NSString *expectedRequestURL =
      [NSString stringWithFormat:@"%@%@", [self.mockAPIService baseURL],
                                 @"/projects/project_id/apps/app_id:exchangeDeviceCheckToken"];
  id URLValidationArg = [OCMArg checkWithBlock:^BOOL(NSURL *URL) {
    XCTAssertEqualObjects(URL.absoluteString, expectedRequestURL);
    return YES;
  }];

  id HTTPBodyValidationArg = [self HTTPBodyValidationArgWithDeviceToken:deviceTokenData
                                                             limitedUse:NO];

  NSData *responseBody =
      [GACFixtureLoader loadFixtureNamed:@"FACTokenExchangeResponseSuccess.json"];
  XCTAssertNotNil(responseBody);

  NSHTTPURLResponse *HTTPResponse = [GACURLSessionOCMockStub HTTPResponseWithCode:200];
  GULURLSessionDataResponse *APIResponse =
      [[GULURLSessionDataResponse alloc] initWithResponse:HTTPResponse HTTPBody:responseBody];

  OCMExpect([self.mockAPIService sendRequestWithURL:URLValidationArg
                                         HTTPMethod:@"POST"
                                               body:HTTPBodyValidationArg
                                  additionalHeaders:@{@"Content-Type" : @"application/json"}])
      .andReturn([FBLPromise resolvedWith:APIResponse]);

  // 1.2. Stub response parsing.
  FBLPromise *rejectedPromise = [FBLPromise pendingPromise];
  [rejectedPromise reject:parsingError];
  OCMExpect([self.mockAPIService appCheckTokenWithAPIResponse:APIResponse])
      .andReturn(rejectedPromise);

  // 2. Send request.
  __auto_type tokenPromise = [self.APIService appCheckTokenWithDeviceToken:deviceTokenData
                                                                limitedUse:NO];

  // 3. Verify.
  XCTAssert(FBLWaitForPromisesWithTimeout(1));

  XCTAssertTrue(tokenPromise.isRejected);
  XCTAssertEqualObjects(tokenPromise.error, parsingError);
  XCTAssertNil(tokenPromise.value);

  OCMVerifyAll(self.mockAPIService);
}

- (void)testAppCheckTokenNetworkError {
  NSData *deviceTokenData = [@"device_token" dataUsingEncoding:NSUTF8StringEncoding];
  NSError *APIError = [NSError errorWithDomain:@"testAppCheckTokenNetworkError"
                                          code:-1
                                      userInfo:nil];

  // 1. Stub API service.
  FBLPromise *rejectedPromise = [FBLPromise pendingPromise];
  [rejectedPromise reject:APIError];

  id HTTPBodyValidationArg = [self HTTPBodyValidationArgWithDeviceToken:deviceTokenData
                                                             limitedUse:NO];
  OCMExpect([self.mockAPIService sendRequestWithURL:[OCMArg any]
                                         HTTPMethod:@"POST"
                                               body:HTTPBodyValidationArg
                                  additionalHeaders:@{@"Content-Type" : @"application/json"}])
      .andReturn(rejectedPromise);

  // 2. Send request.
  __auto_type tokenPromise = [self.APIService appCheckTokenWithDeviceToken:deviceTokenData
                                                                limitedUse:NO];

  // 3. Verify.
  XCTAssert(FBLWaitForPromisesWithTimeout(1));

  XCTAssertTrue(tokenPromise.isRejected);
  XCTAssertNil(tokenPromise.value);
  XCTAssertEqualObjects(tokenPromise.error, APIError);

  OCMVerifyAll(self.mockAPIService);
}

- (void)testAppCheckTokenEmptyDeviceToken {
  NSData *deviceTokenData = [NSData data];

  // 1. Stub API service.
  OCMReject([self.mockAPIService sendRequestWithURL:[OCMArg any]
                                         HTTPMethod:[OCMArg any]
                                               body:[OCMArg any]
                                  additionalHeaders:[OCMArg any]]);

  // 2. Send request.
  __auto_type tokenPromise = [self.APIService appCheckTokenWithDeviceToken:deviceTokenData
                                                                limitedUse:NO];

  // 3. Verify.
  XCTAssert(FBLWaitForPromisesWithTimeout(1));

  XCTAssertTrue(tokenPromise.isRejected);
  XCTAssertNil(tokenPromise.value);

  XCTAssertNotNil(tokenPromise.error);
  XCTAssertEqualObjects(tokenPromise.error.domain, GACAppCheckErrorDomain);
  XCTAssertEqual(tokenPromise.error.code, GACAppCheckErrorCodeUnknown);

  // Expect response body and HTTP status code to be included in the error.
  NSString *failureReason = tokenPromise.error.userInfo[NSLocalizedFailureReasonErrorKey];
  XCTAssertEqualObjects(failureReason, @"DeviceCheck token must not be empty.");

  OCMVerifyAll(self.mockAPIService);
}

#pragma mark - Helpers

- (id)HTTPBodyValidationArgWithDeviceToken:(NSData *)deviceToken limitedUse:(BOOL)limitedUse {
  return [OCMArg checkWithBlock:^BOOL(NSData *body) {
    NSDictionary<NSString *, id> *decodedData = [NSJSONSerialization JSONObjectWithData:body
                                                                                options:0
                                                                                  error:nil];
    XCTAssert([decodedData isKindOfClass:[NSDictionary class]]);

    NSString *base64EncodedDeviceToken = decodedData[@"device_token"];
    XCTAssertNotNil(base64EncodedDeviceToken);

    NSNumber *decodedLimitedUse = decodedData[@"limited_use"];
    XCTAssertNotNil(decodedLimitedUse);
    XCTAssertEqualObjects(decodedLimitedUse, @(limitedUse));

    NSData *decodedToken = [[NSData alloc] initWithBase64EncodedString:base64EncodedDeviceToken
                                                               options:0];
    XCTAssertEqualObjects(decodedToken, deviceToken);
    return YES;
  }];
}

@end

File: AppCheckCore/Tests/Unit/DeviceCheckProvider/GACDeviceCheckAPIServiceTests.swift
import XCTest
#if canImport(Promises)
import Promises
#endif
@testable import AppCheckCore

private class MockAppCheckAPIService: NSObject, _GACAppCheckAPIServiceProtocol {
    var baseURL: String = "https://test.appcheck.url.com/alpha"
    
    var passedRequestURL: URL?
    var passedHTTPMethod: String?
    var passedBody: Data?
    var passedAdditionalHeaders: [String: String]?
    
    var sendRequestResult: Result<_GACURLSessionDataResponse, Error>?
    var appCheckTokenResult: Result<GACAppCheckToken, Error>?
    
    var passedAPIResponse: _GACURLSessionDataResponse?
    
    func sendRequest(with requestURL: URL, httpMethod: String, body: Data?, additionalHeaders: [String : String]?) -> FBLPromise<_GACURLSessionDataResponse> {
        passedRequestURL = requestURL
        passedHTTPMethod = httpMethod
        passedBody = body
        passedAdditionalHeaders = additionalHeaders
        
        let promise = FBLPromise<_GACURLSessionDataResponse>.pending()
        if let result = sendRequestResult {
            switch result {
            case .success(let response): promise.fulfill(response)
            case .failure(let error): promise.reject(error)
            }
        }
        return promise
    }
    
    func appCheckToken(withAPIResponse response: _GACURLSessionDataResponse) -> FBLPromise<GACAppCheckToken> {
        passedAPIResponse = response
        let promise = FBLPromise<GACAppCheckToken>.pending()
        if let result = appCheckTokenResult {
            switch result {
            case .success(let token): promise.fulfill(token)
            case .failure(let error): promise.reject(error)
            }
        }
        return promise
    }
}

class GACDeviceCheckAPIServiceTests: XCTestCase {
    var apiService: GACDeviceCheckAPIService!
    private var mockAPIService: MockAppCheckAPIService!
    
    let kResourceName = "projects/project_id/apps/app_id"
    
    override func setUp() {
        super.setUp()
        mockAPIService = MockAppCheckAPIService()
        apiService = GACDeviceCheckAPIService(apiService: mockAPIService, resourceName: kResourceName)
    }
    
    override func tearDown() {
        apiService = nil
        mockAPIService = nil
        super.tearDown()
    }
    
    func testAppCheckTokenSuccess() async throws {
        try await testAppCheckTokenSuccess(withLimitedUse: false)
    }
    
    func testAppCheckTokenSuccessWithLimitedUse() async throws {
        try await testAppCheckTokenSuccess(withLimitedUse: true)
    }
    
    func testAppCheckTokenSuccess(withLimitedUse limitedUse: Bool) async throws {
        let deviceTokenData = "device_token".data(using: .utf8)!
        let expectedResult = GACAppCheckToken(token: "app_check_token", expirationDate: Date())
        
        let expectedRequestURL = "\(mockAPIService.baseURL)/projects/project_id/apps/app_id:exchangeDeviceCheckToken"
        
        // Since we aren't using the fixture loader for the fake response, we'll just mock any response data.
        let responseBody = "{}".data(using: .utf8)!
        let httpResponse = HTTPURLResponse(url: URL(string: expectedRequestURL)!, statusCode: 200, httpVersion: nil, headerFields: nil)!
        let apiResponse = _GACURLSessionDataResponse(response: httpResponse, httpBody: responseBody)
        
        mockAPIService.sendRequestResult = .success(apiResponse)
        mockAPIService.appCheckTokenResult = .success(expectedResult)
        
        let token = try await apiService.appCheckToken(deviceToken: deviceTokenData, limitedUse: limitedUse)
        
        XCTAssertEqual(token.token, expectedResult.token)
        XCTAssertEqual(token.expirationDate, expectedResult.expirationDate)
        
        XCTAssertEqual(mockAPIService.passedRequestURL?.absoluteString, expectedRequestURL)
        XCTAssertEqual(mockAPIService.passedHTTPMethod, "POST")
        XCTAssertEqual(mockAPIService.passedAdditionalHeaders?["Content-Type"], "application/json")
        try assertHTTPBody(mockAPIService.passedBody, deviceToken: deviceTokenData, limitedUse: limitedUse)
        XCTAssertEqual(mockAPIService.passedAPIResponse, apiResponse)
    }
    
    func testAppCheckTokenResponseParsingError() async throws {
        let deviceTokenData = "device_token".data(using: .utf8)!
        let parsingError = NSError(domain: "testAppCheckTokenResponseParsingError", code: -1, userInfo: nil)
        
        let expectedRequestURL = "\(mockAPIService.baseURL)/projects/project_id/apps/app_id:exchangeDeviceCheckToken"
        let responseBody = "{}".data(using: .utf8)!
        let httpResponse = HTTPURLResponse(url: URL(string: expectedRequestURL)!, statusCode: 200, httpVersion: nil, headerFields: nil)!
        let apiResponse = _GACURLSessionDataResponse(response: httpResponse, httpBody: responseBody)
        
        mockAPIService.sendRequestResult = .success(apiResponse)
        mockAPIService.appCheckTokenResult = .failure(parsingError)
        
        do {
            _ = try await apiService.appCheckToken(deviceToken: deviceTokenData, limitedUse: false)
            XCTFail("Expected error to be thrown")
        } catch let error as NSError {
            XCTAssertEqual(error, parsingError)
        }
        
        XCTAssertEqual(mockAPIService.passedRequestURL?.absoluteString, expectedRequestURL)
        XCTAssertEqual(mockAPIService.passedHTTPMethod, "POST")
        XCTAssertEqual(mockAPIService.passedAdditionalHeaders?["Content-Type"], "application/json")
        try assertHTTPBody(mockAPIService.passedBody, deviceToken: deviceTokenData, limitedUse: false)
        XCTAssertEqual(mockAPIService.passedAPIResponse, apiResponse)
    }
    
    func testAppCheckTokenNetworkError() async throws {
        let deviceTokenData = "device_token".data(using: .utf8)!
        let apiError = NSError(domain: "testAppCheckTokenNetworkError", code: -1, userInfo: nil)
        
        mockAPIService.sendRequestResult = .failure(apiError)
        
        do {
            _ = try await apiService.appCheckToken(deviceToken: deviceTokenData, limitedUse: false)
            XCTFail("Expected error to be thrown")
        } catch let error as NSError {
            XCTAssertEqual(error, apiError)
        }
        
        try assertHTTPBody(mockAPIService.passedBody, deviceToken: deviceTokenData, limitedUse: false)
    }
    
    func testAppCheckTokenEmptyDeviceToken() async throws {
        let deviceTokenData = Data()
        
        do {
            _ = try await apiService.appCheckToken(deviceToken: deviceTokenData, limitedUse: false)
            XCTFail("Expected error to be thrown")
        } catch let error as NSError {
            XCTAssertEqual(error.domain, AppCheckCoreErrorDomain)
            XCTAssertEqual(error.code, AppCheckCoreErrorCode.unknown.rawValue)
            let failureReason = error.userInfo[NSLocalizedFailureReasonErrorKey] as? String
            XCTAssertEqual(failureReason, "DeviceCheck token must not be empty.")
        }
        
        XCTAssertNil(mockAPIService.passedRequestURL)
    }
    
    // MARK: - Helpers
    
    func assertHTTPBody(_ body: Data?, deviceToken: Data, limitedUse: Bool) throws {
        let unwrappedBody = try XCTUnwrap(body)
        let decodedData = try JSONSerialization.jsonObject(with: unwrappedBody, options: []) as? [String: Any]
        let unwrappedDecodedData = try XCTUnwrap(decodedData)
        
        let base64EncodedDeviceToken = try XCTUnwrap(unwrappedDecodedData["device_token"] as? String)
        let decodedLimitedUse = try XCTUnwrap(unwrappedDecodedData["limited_use"] as? Bool)
        
        XCTAssertEqual(decodedLimitedUse, limitedUse)
        
        let decodedToken = try XCTUnwrap(Data(base64Encoded: base64EncodedDeviceToken))
        XCTAssertEqual(decodedToken, deviceToken)
    }
}
